test/callbacks.js

```diff
@@ -0,0 +1,121 @@
+var invokedHooks = [];
+var done = false;
+
+function callback( name ) {
+	return function() {
+		if ( done ) {
+			return;
+		}
+
+		invokedHooks.push( name );
+	};
+}
+
+QUnit.begin( callback( "begin1" ) );
+QUnit.begin( callback( "begin2" ) );
+QUnit.moduleStart( callback( "moduleStart1" ) );
+QUnit.moduleStart( callback( "moduleStart2" ) );
+QUnit.testStart( callback( "testStart1" ) );
+QUnit.testStart( callback( "testStart2" ) );
+
+QUnit.log( callback( "log1" ) );
+QUnit.log( callback( "log2" ) );
+
+// Should the "done" category of callbacks be FIFO or LIFO?
+QUnit.testDone( callback( "testDone1" ) );
+QUnit.testDone( callback( "testDone2" ) );
+QUnit.moduleDone( callback( "moduleDone1" ) );
+QUnit.moduleDone( callback( "moduleDone2" ) );
+QUnit.done( callback( "done1" ) );
+QUnit.done( callback( "done2" ) );
+
+QUnit.done( function() {
+	if ( done ) {
+		return;
+	}
+
+	done = true;
+
+	QUnit.test( "verify callback order", function( assert ) {
+		assert.deepEqual( invokedHooks, [
+			"begin1",
+			"begin2",
+			"moduleStart1",
+			"moduleStart2",
+			"testStart1",
+			"testStart2",
+			"module1 > before",
+			"module1 > beforeEach",
+			"module1 > test1",
+			"log1",
+			"log2",
+			"module1 > afterEach",
+			"testDone1",
+			"testDone2",
+			"moduleDone1", // This shouldn't happen, should it?
+			"moduleDone2", // This shouldn't happen, should it?
```

Comment:
I finally got it. It's interesting to see how the module1's `after` hook isn't triggered, at least.